hypothesis-python/src/hypothesis/ghostwriter.py

```diff
@@ -0,0 +1,192 @@
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+"""
+WARNING: this module is under development and should not be used... yet.
+See https://github.com/HypothesisWorks/hypothesis/pull/2344 for progress.
+"""
+
+import builtins
+import enum
+import inspect
+from collections import OrderedDict
+from textwrap import indent
+from typing import Callable, Dict, Tuple, Type, Union
+
+import hypothesis.strategies as st
+from hypothesis.errors import InvalidArgument
+from hypothesis.utils.conventions import InferType, infer
+
+IMPORT_SECTION = """
+# This test template was produced by the `hypothesis.ghostwriter` module.
+{imports}
+"""
+
+TEMPLATE = """
+{nothing_hint}@given({given_args})
+def test_{test_kind}_{func_name}({arg_names}):
+{test_body}
+"""
+
+SUPPRESS_BLOCK = """\
+try:
+{test_body}
+except {exceptions}:
+    assume(False)
+"""
+
+Except = Union[Type[Exception], Tuple[Type[Exception], ...]]
+
+
+def _check_except(except_: Except) -> Tuple[Type[Exception], ...]:
+    if isinstance(except_, tuple):
+        for i, e in enumerate(except_):
+            if not isinstance(e, type) or not issubclass(e, Exception):
+                raise InvalidArgument(
+                    "Expected an Exception but got except_[%s]=%r (type=%s)"
+                    % (i, e, type(e).__name__)
+                )
+        return except_
+    if not isinstance(except_, type) or not issubclass(except_, Exception):
+        raise InvalidArgument(
+            "Expected an Exception or tuple of exceptions, but got except_=%r (type=%s)"
+            % (except_, type(except_).__name__)
+        )
+    return (except_,)
+
+
+def _strategy_for(param: inspect.Parameter) -> Union[st.SearchStrategy, InferType]:
+    # We use `infer` and go via `builds()` instead of directly through
+    # `from_type()` so that `get_type_hints()` can resolve any forward
+    # references for us.
+    if param.annotation is not inspect.Parameter.empty:
+        return infer
+    # If there's no annotation and no default value, the user will have to
+    # fill this in later.  We use nothing() as a placeholder to this effect.
+    if param.default is inspect.Parameter.empty:
+        return st.nothing()
+    # If our default value is an Enum or a boolean, we assume that any value
+    # of that type is acceptable.  Otherwise, we only generate the default.
+    if isinstance(param.default, bool):
+        return st.booleans()
+    if isinstance(param.default, enum.Enum):
+        return st.sampled_from(type(param.default))
+    return st.just(param.default)
```

Comment:
Doesn't have to be done as part of this PR, but I'm vaguely concerned with the fact that `param.default` may not have a particularly legible text representation. In particular it might not be valid Python code. Could/should we check for that and return `st.nothing()` as part of this?

Reply:
Hmm.  I think you're right, syntatically-invalid-code from reprs could be an issue... if you don't mind I'll leave that for the next PR, though - the right place to handle it is probably when repr-ing the strategies.

I expect that to involve switching to `st.nothing()`, but also dropping in an explanatory comment (or replacing the generic one) and possibly improving some of our strategy repr logic to e.g. use qualnames if the repr is invalid code.

Reply:
Yup, happy to handle this in the next PR. Can we add a note of it somewhere so it doesn't get lost?

RE qualnames: I think that may already be handled. Note that we use the pretty printer library for this rather than straight up reprs (I think!)